File: grooves-api/src/main/groovy/com/github/rahulsom/grooves/api/snapshots/internal/BaseJoin.java
package com.github.rahulsom.grooves.api.snapshots.internal;

import com.github.rahulsom.grooves.api.AggregateType;
import com.github.rahulsom.grooves.api.events.BaseEvent;

import java.util.List;

/**
 * A special kind of {@link BaseSnapshot} that stores information about joined entities.
 *
 * @param <AggregateT>         The Aggregate this join represents
 * @param <JoinIdT>            The type for the join's {@link #getId()} field
 * @param <JoinedAggregateIdT> The type for the other aggregate that {@link AggregateT} joins to
 * @param <EventIdT>           The type for the {@link EventT}'s id field
 * @param <EventT>             The base type for events that apply to {@link AggregateT}
 *
 * @author Rahul Somasunderam
 */
public interface BaseJoin<
        AggregateT extends AggregateType,
        JoinIdT,
        JoinedAggregateIdT,
        EventIdT,
        EventT extends BaseEvent<AggregateT, EventIdT, EventT>
        >
        extends BaseSnapshot<AggregateT, JoinIdT, EventIdT, EventT> {

    List<JoinedAggregateIdT> getJoinedIds();

    void setJoinedIds(List<JoinedAggregateIdT> ids);

}

File: grooves-types/src/main/kotlin/com/github/rahulsom/grooves/api/snapshots/internal/BaseJoin.kt
package com.github.rahulsom.grooves.api.snapshots.internal

import com.github.rahulsom.grooves.api.events.BaseEvent

/**
 * A special kind of [BaseSnapshot] that stores information about joined entities.
 *
 * @param [AggregateT] The Aggregate this snapshot works over
 * @param [JoinIdT] The type for [BaseJoin.id]
 * @param [EventIdT] The type for [BaseEvent.id]
 * @param [EventT] The base type for events that apply to [AggregateT]
 *
 * @author Rahul Somasunderam
 */
interface BaseJoin<
        AggregateT,
        JoinIdT,
        in JoinedAggregateT,
        EventIdT,
        in EventT : BaseEvent<AggregateT, EventIdT, in EventT>> :
    BaseSnapshot<AggregateT, JoinIdT, EventIdT, EventT> {

    fun addJoinedAggregate(joinedAggregateT: JoinedAggregateT)
    fun removeJoinedAggregate(joinedAggregateT: JoinedAggregateT)

}
